tests/src/test_cases/test_btnmatrix.c

```diff
@@ -0,0 +1,390 @@
+{
+    static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
+    lv_btnmatrix_set_map(btnm, btn_map);
+
+    const char * lv_btnmatrix_get_btn_text(const lv_obj_t * obj, uint16_t btn_id);
+
+    TEST_ASSERT_EQUAL_STRING("A", lv_btnmatrix_get_btn_text(btnm, 0));
+    TEST_ASSERT_EQUAL_STRING("B", lv_btnmatrix_get_btn_text(btnm, 1));
+    TEST_ASSERT_EQUAL_STRING("C", lv_btnmatrix_get_btn_text(btnm, 2));
+    TEST_ASSERT_EQUAL_STRING("D", lv_btnmatrix_get_btn_text(btnm, 3));
+}
+
+/* Common event handler for all the consecutive test cases. */
+static void event_handler(lv_event_t * e)
+{
+    lv_event_code_t code = lv_event_get_code(e);
+    if(code == exp_evt_code) {
+        event_triggered = true;
+    }
+}
+
+void test_btn_matrix_pressed_event_works(void)
+{
+    static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
+    lv_btnmatrix_set_map(btnm, btn_map);
+    lv_obj_add_event_cb(btnm, event_handler, LV_EVENT_PRESSED, NULL);
+    /* Set expected event code before the event is raised. */
+    exp_evt_code = LV_EVENT_PRESSED;
+    /* Click button index 0. */
+    lv_test_mouse_click_at(10, 10);
+    TEST_ASSERT_TRUE(event_triggered);
+}
+
+void test_btn_matrix_release_event_works(void)
+{
+    static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
+    lv_btnmatrix_set_map(btnm, btn_map);
+    lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
+    lv_obj_add_event_cb(btnm, event_handler, LV_EVENT_RELEASED, NULL);
+    /* Set expected event code before the event is raised. */
+    exp_evt_code = LV_EVENT_RELEASED;
+
+    /* Click button index 0. */
+    lv_test_mouse_click_at(65, 35);
+    TEST_ASSERT_TRUE(event_triggered);
+    event_triggered = false;
+    /* This will increase test coverage by unchecking the
+     same button. */
+    lv_btnmatrix_set_btn_ctrl(btnm, 0, LV_BTNMATRIX_CTRL_POPOVER);
+    /* Click button index 0. */
+    lv_test_mouse_click_at(65, 35);
+    TEST_ASSERT_TRUE(event_triggered);
+}
+
+void test_btn_matrix_size_changed_event_works(void)
+{
+    static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
+    lv_btnmatrix_set_map(btnm, btn_map);
+    lv_obj_add_event_cb(btnm, event_handler, LV_EVENT_SIZE_CHANGED, NULL);
+    /* Set expected event code before the event is raised. */
+    exp_evt_code = LV_EVENT_SIZE_CHANGED;
+    lv_event_send(btnm, LV_EVENT_SIZE_CHANGED, NULL);
+    TEST_ASSERT_TRUE(event_triggered);
+}
+
+void test_btn_matrix_key_event_works(void)
+{
+    char keyCode;
+    static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
+    lv_btnmatrix_set_map(btnm, btn_map);
+    lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
+    lv_obj_add_event_cb(btnm, event_handler, LV_EVENT_KEY, NULL);
+    /* Set expected event code before the event is raised. */
+    exp_evt_code = LV_EVENT_KEY;
+
+    keyCode = LV_KEY_RIGHT;
+    lv_event_send(btnm, LV_EVENT_KEY, &keyCode);
+    TEST_ASSERT_TRUE(event_triggered);
```

Comment:
Please also test the  `selected_btn` here.

Reply:
Can you please show me how to capture the reference image from the simulator?

Reply:
I'm on Linux Mint and the built in screenshot utility can capture a window without border. 

On Windows you can press print screen and clip to the window content using Paint. 

Don't forget to use `LV_COLOR_DEPTH 32` and 800x480 resolution.

Reply:
I'm running the simulator on WSL. The screen resolution seems to be correct. But I still get an image mismatch error.
<img width="399" alt="btnm_3" src="https://user-images.githubusercontent.com/46624548/188200066-b93e4143-765d-4530-a042-cff197dd0346.png">

In the test_screenshot_error.h header I see the comment `//Diff in btnm_3.png at (0;0), f5f5f5 instead of e3e3e3)`.
Seems to be a color mismatch.

Reply:
I have attached the header file for your reference. Can you please check it?
[Uploading test_screenshot_error.txt…]()

Reply:
I have attached the header file for your reference. Can you please check it?
[test_screenshot_error.txt](https://github.com/lvgl/lvgl/files/9479785/test_screenshot_error.txt)

Reply:
I see an extra pixel border on the top and right of @mmurty's screenshot which is probably the cause of the error.

Ideally there should be a way of autogenerating the reference image if one doesn't exist; maybe I'll look into this if I find some time.

Reply:
Earlier I captured the image using windows snipping tool. Now I have captured the image using printscreen and clipped it using paint. Now the color difference is not much but it still fails. 
Attaching the files for your reference. 
[test_screenshot_error.txt](https://github.com/lvgl/lvgl/files/9479901/test_screenshot_error.txt)
<img width="400" alt="btnm_4" src="https://user-images.githubusercontent.com/46624548/188204750-ddaf4718-03a4-4d12-8cb4-3cb4a37c7800.png">

Reply:
Are you sure the `LV_COLOR_DEPTH` is 32?

Reply:
I'm not sure about the tests but the simulator has `#define LV_COLOR_DEPTH     32` defined in the lv_conf.h file. 
I'm running the simulator on WSL + VSCode.

Reply:
Please try with this image:
![image](https://user-images.githubusercontent.com/7599318/188381195-6b0caa69-df56-409d-bf24-48a7f8e4a9e6.png)

Although the first pixel pixels are f5f5f5 here too. It seems the tests calculate different bg color. Please add this to the test:
```c
  printf("Screen bg color: %x\n",lv_obj_get_style_bg_color(lv_scr_act(), 0));
```
It prints 0xfff5f5f5 for me.

Reply:
The test passes with the above image. It seems as I am cropping the image manually it doesn't come exactly 800x480 pixels in size and hence it doesn't match with the framebuffer.

Reply:
We will figure out somethings to make it simpler but for now, please copy the code you are planning to use for testing and I'll send back the image.

Reply:
In the commited code I see that the test passes for left and right keys but fails for up and down keys.
I am testing in the simulator and the reported key index is correct. 
But in the `LV_KEY_DOWN` event, when the below code executes,
```
for(area_below = btnm->btn_id_sel; area_below < btnm->btn_cnt; area_below++) {
                    if(btnm->button_areas[area_below].y1 > btnm->button_areas[btnm->btn_id_sel].y1 &&
                       pr_center >= btnm->button_areas[area_below].x1 &&
                       pr_center <= btnm->button_areas[area_below].x2 + col_gap &&
                       button_is_inactive(btnm->ctrl_bits[area_below]) == false &&
                       button_is_hidden(btnm->ctrl_bits[area_below]) == false) {
                        break;
                    }
```
I added following prints in the code:
```
printf("Area below = %d\n", area_below);
printf("Selected BtnId = %d\n", btnm->btn_id_sel);
printf("Area below Y1 = %d\n", btnm->button_areas[area_below].y1);
printf("Selected Y1 = %d\n", btnm->button_areas[btnm->btn_id_sel].y1);
```
I notice that Y1 value for the buttons of second row is -2147483645. 
Anything you can remember that I'm missing?

Reply:
Hello,
Please provide the screenshot for the following code:
```
static lv_obj_t * active_screen = NULL;

active_screen = lv_scr_act();
btnm = lv_btnmatrix_create(active_screen);
  
static const char * btn_map[] = {"A", "B", "\n", "C", "D", ""};
lv_btnmatrix_set_map(btnm, btn_map);
lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
lv_btnmatrix_set_btn_ctrl(btnm, 0, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 3, LV_BTNMATRIX_CTRL_CHECKED);
//lv_obj_add_event_cb(btnm, event_handler, LV_EVENT_KEY, NULL);
lv_obj_set_pos(btnm, 10, 10);

static const char * btn_map2[] = {"A", "\n", "B", "\n", "C", "\n", "D", ""};
btnm = lv_btnmatrix_create(active_screen);
lv_obj_set_width(btnm, 150);
lv_obj_set_height(btnm, 250);
lv_btnmatrix_set_map(btnm, btn_map2);
lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
lv_btnmatrix_set_one_checked(btnm, true);
lv_btnmatrix_set_btn_ctrl(btnm, 1, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 2, LV_BTNMATRIX_CTRL_CHECKED);
lv_obj_set_pos(btnm, 10, 160);

static const char * btn_map3[] = {"A", "B", "C", "\n", "D", "E", "F", ""};
btnm = lv_btnmatrix_create(active_screen);
lv_btnmatrix_set_map(btnm, btn_map3);
lv_obj_set_width(btnm, 400);
lv_btnmatrix_set_btn_width(btnm, 0, 3);
lv_btnmatrix_set_btn_width(btnm, 1, 2);
lv_btnmatrix_set_btn_width(btnm, 2, 1);
lv_btnmatrix_set_btn_width(btnm, 3, 1);
lv_btnmatrix_set_btn_width(btnm, 4, 2);
lv_btnmatrix_set_btn_width(btnm, 5, 3);
lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
lv_btnmatrix_set_btn_ctrl(btnm, 1, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 4, LV_BTNMATRIX_CTRL_CHECKED);
lv_obj_set_pos(btnm, 300, 10);

static const char * btn_map4[] = {"A", "B", "C", "D", "E", "F", "G", "\n", 
                                  "G", "F", "E", "D", "C", "B", "A", ""};
btnm = lv_btnmatrix_create(active_screen);
lv_btnmatrix_set_map(btnm, btn_map4);
lv_obj_set_width(btnm, 600);
lv_obj_set_height(btnm, 150);
lv_btnmatrix_set_btn_width(btnm, 0, 1);
lv_btnmatrix_set_btn_width(btnm, 1, 2);
lv_btnmatrix_set_btn_width(btnm, 2, 3);
lv_btnmatrix_set_btn_width(btnm, 3, 4);
lv_btnmatrix_set_btn_width(btnm, 4, 5);
lv_btnmatrix_set_btn_width(btnm, 5, 6);
lv_btnmatrix_set_btn_width(btnm, 6, 7);

lv_btnmatrix_set_btn_width(btnm, 7, 7);
lv_btnmatrix_set_btn_width(btnm, 8, 6);
lv_btnmatrix_set_btn_width(btnm, 9, 5);
lv_btnmatrix_set_btn_width(btnm, 10, 4);
lv_btnmatrix_set_btn_width(btnm, 11, 3);
lv_btnmatrix_set_btn_width(btnm, 12, 2);
lv_btnmatrix_set_btn_width(btnm, 13, 1);

lv_btnmatrix_set_btn_ctrl_all(btnm, LV_BTNMATRIX_CTRL_CHECKABLE);
lv_btnmatrix_set_btn_ctrl(btnm, 1, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 3, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 5, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 7, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 9, LV_BTNMATRIX_CTRL_CHECKED);
lv_btnmatrix_set_btn_ctrl(btnm, 11, LV_BTNMATRIX_CTRL_CHECKED);
lv_obj_set_pos(btnm, 180, 160);
```

Reply:
Here is the screenshot:
![image](https://user-images.githubusercontent.com/7599318/188826861-79a36e52-c433-4c1e-8c1f-03632dca2f4c.png)


How can I reproduce the KEY_DOWN issue?

Reply:
I have committed the failing test case. So you can have a look at that.

Reply:
Hi,
Did you get a chance to look at the failure?

Reply:
I've just tried it out and found the issue too. `lv_obj_update_layout(btnm);` needs to be added after setting the map and ctrl map. 

It's required because LVGL calculates the actual coordinates of the object only before rendering. So when the btnmatrix was created it had 0;0 size. Therefore the button areas were messed up too. 

When the size of the widget is set in `lv_obj_update_layout`, the area of the buttons is recalculated. 

I pushed the fix. If the CI passed I merge the PR.

Reply:
Thanks.
There are still some minor tests which are to be added. (Some basic render tests which you suggested earlier).
So should I raise a new PR for that?

Reply:
Just please add the rendering test for which I send the screenshot, and I think it's ready.

Reply:
I have added the rendering test. Please check.

Reply:
Thanks! CI, please check :)